### PR TITLE
8314906: [testbug] Create behavior tests for text input controls

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Creating the first batch of tests and testing framework that enables writing behavior tests for javafx controls, focusing on key bindings.  The idea is to make writing such tests a simple process.  

This PR deals with the descendants of TextInputControl (TextField, PasswordField, TextArea).  Most of the tests are headless, but in some cases (TextArea) a headful test is required because the behavior needs rendered text to function (example: page up / page down / line start / line end and the like).

The tests exercise the key bindings registered by the Skin (or, rather the associated Behavior) at least once, and sometimes more than once.

Some mappings cannot be tested due to Robot not supporting keypad events (created [JDK-8316307](https://bugs.openjdk.org/browse/JDK-8316307)).

In addition, the key bindings are documented in /doc-files/behavior markdown documents:

https://github.com/andy-goryachev-oracle/jfx/blob/8314906.behavior.test/doc-files/behavior/PasswordField.md
https://github.com/andy-goryachev-oracle/jfx/blob/8314906.behavior.test/doc-files/behavior/TextArea.md
https://github.com/andy-goryachev-oracle/jfx/blob/8314906.behavior.test/doc-files/behavior/TextField.md
https://github.com/andy-goryachev-oracle/jfx/blob/8314906.behavior.test/doc-files/behavior/TextInputControl.md

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8314906](https://bugs.openjdk.org/browse/JDK-8314906): [testbug] Create behavior tests for text input controls (**Bug** - P4)

### Reviewers
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Karthik P K](https://openjdk.org/census#kpk) (@karthikpandelu - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1221/head:pull/1221` \
`$ git checkout pull/1221`

Update a local copy of the PR: \
`$ git checkout pull/1221` \
`$ git pull https://git.openjdk.org/jfx.git pull/1221/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1221`

View PR using the GUI difftool: \
`$ git pr show -t 1221`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1221.diff">https://git.openjdk.org/jfx/pull/1221.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1221#issuecomment-1719925410)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
This is a large enough set of tests that a second pair of eyes will be useful.

/reviewers 2

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/pgstub/StubToolkit.java [hunk 1]
```diff
@@ -929,4 +959,14 @@ public List<MenuBase> getMenus() {
...
+
+    /**
+     * Ensures that the current (test) runs with the StubToolkit.
+     * Throws a class cast exception otherwise.
+     *
+     * @return the StubToolkit instance
+     */
+    public static StubToolkit ensure() {
+        return (StubToolkit)Toolkit.getToolkit();
+    }
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/pgstub/StubToolkit.java:971]: This is unused (and may not be needed). I think you can revert the changes to this file.

File: tests/system/src/test/java/test/util/Util.java [hunk 2]
```diff
@@ -466,4 +471,28 @@ public static boolean isFractionalScaleX(Node n) {
     private static boolean isFractional(double x) {
         return x != Math.rint(x);
     }
+
+    /**
+     * Returns true if the operating system is a form of Windows.
+     * @return true if the platform is Windows
+     */
+    public static boolean isWindows(){
+        return WINDOWS;
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/util/Util.java:480]: Are these really needed or can the tests call PlatformUtil directly? If there is a good reason to add them, then they should delegate to PlatformUtil rather than duplicate the functionality.

**Review by kevinrushforth — COMMENTED** [review]
A couple quick comments I spotted while skimming this PR.

File: tests/system/src/test/java/test/javafx/scene/control/behavior/Mod.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ */
+package test.javafx.scene.control.behavior;
+
+import javafx.scene.input.KeyCode;
+import test.util.Util;
+
+/**
+ * Key Modifiers for use in behavior tests.
+ */
+public enum Mod {
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/Mod.java:33]: the obvious would be KeyModifier, but we already have one in test.com.sun.javafx.scene.control.infrastructure with a different and less convenient semantics.

File: tests/system/src/test/java/test/util/Util.java (hunk 2 above)
Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/util/Util.java:480]: was able to use PlatformUtils directly after adding 
--add-exports javafx.base/com.sun.javafx=ALL-UNNAMED
to the launch configuration.

File: tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+                "2", KeyCode.DIGIT2,
+                "3", KeyCode.DIGIT3,
+                "4", KeyCode.DIGIT4,
+                "5", KeyCode.DIGIT5,
+                "6", KeyCode.DIGIT6,
+                "7", KeyCode.DIGIT7,
+                "8", KeyCode.DIGIT8,
+                "9", KeyCode.DIGIT9,
+                ".", KeyCode.PERIOD,
+                ",", KeyCode.COMMA
```
Comment (beldenfox) [inline tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java:196]: On French layouts the DIGIT codes don't actually produce digits (you have to hold down Shift) and there's no KeyCode.PERIOD (also a shifted character). You don't seem to be using those characters but you might want to remove them from this table so no one is tempted to use them in the future.

Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java:196]: Interesting.  

It is my understanding though that for the purposes of this test it should work as expected.  I've added `TextAreaBehaviorRobotTest.testTyping()` - could you try running it with your setup please?

```
    @Test
    public void testTyping() throws Exception {
        execute(
            //addKeyListener(),
            "0123456789.,abracadabra",
            checkText("0123456789.,abracadabra", 23)
        );
    }
```

Comment (beldenfox) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java:196]: I ran the TextAreaBehaviorRobotTest on macOS 13.5.2. Test passed with my keyboard set to U.S. English and failed when set to French.

```
TextAreaBehaviorRobotTest > testTyping() FAILED
    org.opentest4j.AssertionFailedError: in step 2 ==> expected: <0123456789.,abracadabra> but was: <à&é"'(§è!ç,abracadabra>
```

Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java:196]: Merci, Monsieur Fox!

We seem to have quite a few tickets in this area (who created the first one in the list?):

* [JDK-8278924](https://bugs.openjdk.org/browse/JDK-8278924) [Linux] Robot key test can fail if multiple keyboard layouts are installed
* [JDK-8022079](https://bugs.openjdk.org/browse/JDK-8022079) [macosx] swing shortcuts are counterintuitive with non-english keyboard layouts
* [JDK-8087915](https://bugs.openjdk.org/browse/JDK-8087915) Mac: accelerator doesn't take into account azerty keyboard layout
* [JDK-4760902](https://bugs.openjdk.org/browse/JDK-4760902) java.awt.Robot can't generate % character on FRENCH keyboard

Not only we get key events for wrong characters, but also shortcut(A) generates `character = q, text = , code = UNDEFINED, metaDown, shortcutDown]` which closes the test app :-(

I have no idea how to fix this test...  We could try to introduce a translation layer to emit the right sequences (for each keyboard layout), or, alternatively, we could skip the tests altogether for non-US keyboard, except ... there seems to be no easy way to determine the keyboard layout?  (There are some suggestions involving InputContext or JNI on stackoverflow).  We could also try to skip the test if Locale.getDefault is other than Locale.US, not sure if that would solve the issue.

What do you think?

Comment (beldenfox) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java:196]: The problem you're running into here isn't due to any bug in JavaFX. The way KeyCodes identify keys works well for defining accelerators but not for using a Robot to generate a specific printable character. As an extreme example, if you're using a Greek keyboard there will be a key that has KeyCode.Z assigned to it so you can use Shortcut+Z to invoke Undo. But that key will generate a Greek letter when pressed by a human or a Robot.

You can probably assume that anyone running these tests is using a Latin/Roman keyboard that can generate the lower-case letters a-z so one approach is to just stick to those characters in the typing tests. The other is to try to identify the layout using a Robot. For example, send KeyCode.DIGIT1 and see if that generates a '1'. If not you're probably on a French layout.

Punctation and symbols are hopeless, there's just too much variation in the keyboard layouts.

> Not only we get key events for wrong characters, but also shortcut(A) generates `character = q, text = , code = UNDEFINED, metaDown, shortcutDown]` which closes the test app :-(

Could you provide more details (platform, keyboard, keyboard layout, etc.)? On the Mac PR #425 cleaned all this up in jfx21.

Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java:196]: I am running this on macOS 13.5.1 with several keyboard layouts enabled.

A test that does keyPress(KecCode.A) followed by a keyRelease(KeyCode.A) generates the following sequence of events:

With the US layout:

```
KeyEvent [source = TextArea@7da926e[styleClass=text-input text-area], target = TextArea@7da926e[styleClass=text-input text-area], eventType = KEY_PRESSED, consumed = false, character = , text = a, code = A]
KeyEvent [source = TextArea@7da926e[styleClass=text-input text-area], target = TextArea@7da926e[styleClass=text-input text-area], eventType = KEY_TYPED, consumed = false, character = a, text = , code = UNDEFINED]
KeyEvent [source = TextArea@7da926e[styleClass=text-input text-area], target = TextArea@7da926e[styleClass=text-input text-area], eventType = KEY_RELEASED, consumed = false, character = , text = a, code = A]
```

With the French layout:

```
KeyEvent [source = TextArea@75ffb50a[styleClass=text-input text-area], target = TextArea@75ffb50a[styleClass=text-input text-area], eventType = KEY_PRESSED, consumed = false, character = , text = q, code = A]
KeyEvent [source = TextArea@75ffb50a[styleClass=text-input text-area], target = TextArea@75ffb50a[styleClass=text-input text-area], eventType = KEY_TYPED, consumed = false, character = q, text = , code = UNDEFINED]
KeyEvent [source = TextArea@75ffb50a[styleClass=text-input text-area], target = TextArea@75ffb50a[styleClass=text-input text-area], eventType = KEY_RELEASED, consumed = false, character = , text = q, code = A]
```

And with a cyrillic layout it generates whatever cyrillic letter is mapped to the key.  So, basically, KeyCode is more like a scan code, which means there is no reliable way to test key bindings using Robot, unless we limit ourselves (and the test) to the US layout.  Or is there another way?

Screenshot 2023-09-18 at 13 19 51

Comment (beldenfox) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java:196]: This is the behavior we saw on the Mac before PR #425 was integrated. As of JavaFX 21 keyPress(KeyCode.A) should generate an 'a' and never a 'q' on all Latin layouts and all platforms (on a Cyrillic layout you'll still get a Cyrillic character).

I can't reproduce this with the test app (tests/manual/events/KeyboardTest.java). Is there a test in this PR I should be trying?

Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java:196]: You could try this code (using this PR's code):

```
public class KeyboardLayoutTest extends TextInputBehaviorRobotTest<TextArea> {

    public KeyboardLayoutTest() {
        super(new TextArea());
    }

    @BeforeEach
    @Override
    public void beforeEach() {
        super.beforeEach();
        control.setWrapText(true);
    }

    @Test
    public void testTyping() throws Exception {
        KeyCode[] codes = {
            KeyCode.A
        };

        execute(
            exe(() -> {
                control.addEventFilter(KeyEvent.ANY, (ev) -> {
                    System.out.println(ev);
                    ev.consume();
                });
            })
        );

        for(KeyCode cd: codes) {
            Util.runAndWait(() -> {
                robot.keyPress(cd);
                robot.keyRelease(cd);
            });
        }
    }
}
```

Comment (beldenfox) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java:196]: I tried your test code and couldn't reproduce the problem. Both U.S. English and French layouts gave me the expected result (an 'a' character). I also tried a Russian keyboard just to verify that it would generate a Cyrillic character which it did. I'm running the tests against JavaFX built using the sources in your pull request.

French

```
KeyEvent [source = TextArea@7237703[styleClass=text-input text-area], target = TextArea@7237703[styleClass=text-input text-area], eventType = KEY_PRESSED, consumed = false, character = ?, text = a, code = A]
KeyEvent [source = TextArea@7237703[styleClass=text-input text-area], target = TextArea@7237703[styleClass=text-input text-area], eventType = KEY_TYPED, consumed = false, character = a, text = , code = UNDEFINED]
KeyEvent [source = TextArea@7237703[styleClass=text-input text-area], target = TextArea@7237703[styleClass=text-input text-area], eventType = KEY_RELEASED, consumed = false, character = ?, text = a, code = A]
```

U.S. English

```
KeyEvent [source = TextArea@73459ba5[styleClass=text-input text-area], target = TextArea@73459ba5[styleClass=text-input text-area], eventType = KEY_PRESSED, consumed = false, character = ?, text = a, code = A]
KeyEvent [source = TextArea@73459ba5[styleClass=text-input text-area], target = TextArea@73459ba5[styleClass=text-input text-area], eventType = KEY_TYPED, consumed = false, character = a, text = , code = UNDEFINED]
KeyEvent [source = TextArea@73459ba5[styleClass=text-input text-area], target = TextArea@73459ba5[styleClass=text-input text-area], eventType = KEY_RELEASED, consumed = false, character = ?, text = a, code = A]
```

Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java]: Thank you for your help!  I can confirm that the latest code does indeed generate the right key event.

I suspect there was something in my setup that got corrupted - I had to install the latest eclipse, found out it's pretty much unusable due to https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/309 , had to revert back to 2023-06 just to discover that my macOS happily suggested that 'your application is damaged and can't be opened' and do I want to move it to Trash?  So after some xattr'ing, restoring proxy settings, etc., etc. I am back online and the issue is gone.

I did remove the offending characters and modified the tests, so thank you once again - your help was invaluable.

File: modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/KeyEventFirer.java [hunk 5]
```diff
@@ -25,24 +25,28 @@
...
 
-
+/**
+ * Keyboard simulator for use with tests.
+ */
 public class KeyEventFirer {
-
     private final EventTarget target;
     private final Scene scene;
+    private static HashMap<Character,KeyCode> keyCodes;
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/KeyEventFirer.java:49]: minor : space after `,`

File: modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/KeyEventFirer.java [hunk 6]
```diff
@@ -135,18 +233,153 @@ private KeyEvent[] createMirroredEvents(KeyCode keyCode, KeyModifier... modifier
...
+        KeyCode code = keyCodes.get(c);
+        if (code == null) {
+            throw new RuntimeException(String.format("character 0x%04x has no corresponding KeyCode", (int)c));
+        }
+        return code;
+    }
+
+    private static HashMap<Character, KeyCode> createKeyCodes(Object ... pairs) {
+        HashMap<Character, KeyCode> m = new HashMap<>();
+        for(int i=0; i<pairs.length; ) {
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/KeyEventFirer.java:378]: Minor : space before and after `=` and `<`

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/behavior/BehaviorTestBase.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+    /**
+     * Must be called in each test's <code>&#x40;AfterEach</code> method:
+     * <pre>
+     *     &#x40;AfterEach
+     *     public void afterEach() {
+     *         closeStage();
+     *     }
+     * <pre>
+     * @param control the control being tested
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/behavior/BehaviorTestBase.java:82]: Remove `@param` line

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/behavior/PasswordFieldBehaviorTest.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    @Test
+    @Override
+    public void testConsumeEnter() {
+        super.testConsumeEnter();
+    }
+
+    @Test
+    @Override
+    public void testCopy() {
+        // copy is disabled
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/behavior/PasswordFieldBehaviorTest.java:63]: An empty test case simply gets reported as passed.
Do you want to assert that the "copy is disabled" somehow?
There are a few other test cases in this file which are of similar nature.
If they need to be worked upon in future, we can mark them as TODO.
If they are just for information, we can keep the entire test cases commented so that they do not execute and report as passed.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/behavior/TextAreaBehaviorTest.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    @Test
+    @Override
+    public void testConsume() {
+        super.testConsume();
+    }
+
+    @Test
+    @Override
+    public void testConsumeEnter() {
+        // does not consume ENTER
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/behavior/TextAreaBehaviorTest.java:64]: Similar comment about empty test cases in this file.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/behavior/PasswordFieldBehaviorTest.java (hunk 8 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/behavior/PasswordFieldBehaviorTest.java:63]: good point, thanks!

**Comment by andy-goryachev-oracle** [comment]
@karthikpandelu could you also take a look at this please?

**Review by karthikpandelu — CHANGES_REQUESTED** [review]
The tests look good to me.
Left few minor comments inline.

File: doc-files/behavior/PasswordField.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
+# PasswordField Behavior
+
+## Key Binginds
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply doc-files/behavior/PasswordField.md:3]: The first rule of submitting scientific publications for a review: make an obvious spelling error in the title.
:-)

thanks!

**Review by karthikpandelu — APPROVED** [review]
LGTM

**Review by aghaisas — APPROVED** [review]
